### ISSUE TITLE
Bazel rules should use the python headers from the python toolchain

### ISSUE THREAD

**Issue by dhalperi** [grpc/grpc#24665]
### Is your feature request related to a problem? Please describe.
The recommend way to use Python with Bazel is by providing a [Python toolchain](https://github.com/bazelbuild/bazel/blob/master/tools/python/toolchain.bzl#L65).

gRPC's `cython_library` rule, however, uses the [system Python headers](https://github.com/grpc/grpc/blob/a951c7632a6a52c8dc7920105d0dd7295b795469/bazel/cython_library.bzl#L61) instead of the headers for the Python toolchain that is being used.

### Describe the solution you'd like
Use the headers for the Python toolchain selected for the given platform.

### Describe alternatives you've considered
The only alternative I've found is forcing users to keep their system environment in sync with whatever the Bazel Python toolchain they're using is. This, of course, defeats the purpose of adding a Bazel toolchain - isolation from the system environment!

### Additional context
Add any other context about the feature request here.

**Comment by dhalperi** [comment]
This is stale because no one has engaged. I would love a response from the gRPC team :)

**Comment by aaliddell** [comment]
This almost works and I think could be fixed with a one-line change, at least in the short-term.

Presently, the `python_configure` is run which autodiscovers the local Python, then the headers are bound to the name `python_headers`:

https://github.com/grpc/grpc/blob/afc040b07a8dcab704fdcab248c8f8b764fafd6c/bazel/grpc_python_deps.bzl#L56-L61

However, when Cython is run it is linked against the non-aliased headers:

https://github.com/grpc/grpc/blob/afc040b07a8dcab704fdcab248c8f8b764fafd6c/bazel/cython_library.bzl#L71-L76

Therefore, the proposed change would be for `pyx_library` to instead use the binding `@python_headers`, thereby allowing a user to override the binding of `python_headers` in their WORKSPACE to point to the correct headers location. This is the same as how protobuf works when building the C extension: https://github.com/protocolbuffers/protobuf/blob/master/util/python/BUILD

**Comment by adam-azarchs** [comment]
One problem with that solution is that it perpetuates the problems inherent `bind`.

Another more serious problem is that it still won't work if, as in for example my case, the system does not have a version of python installed that matches the version from your toolchain.  For example our python toolchains are configured for python 3.8, so `local_config_python` fails with a message like
```
Error in fail: Python Configuration Error: Unable to find Python headers for /usr/bin/python3
Traceback (most recent call last):
  File "<string>", line 1, in <module>
AssertionError: /usr/include/python3.8/Python.h does not exist.
Are Python headers installed? Try installing python-dev or python3-dev on Debian-based systems. Try python-devel or python3-devel on Redhat-based systems.
```

**Comment by psigen** [comment]
Has anyone found a workaround for this?

Is the fundamental problem that the interpreter toolchain definition in `rules_python` needs to include a target for a header path as well for native compilation purposes?

**Comment by aaliddell** [comment]
Not stale

**Comment by aaliddell** [comment]
The `rules_python` toolchains [now provide](https://github.com/bazelbuild/rules_python/pull/666) the `python_headers` target as part of all of their default toolchains, which has removed the blocker on transitioning from `local_config_python` to fetching the headers from the Python toolchain. Could this be reconsidered now @gnossen?

**Comment by gnossen** [comment]
Yes. This is something I'm actively working on. After a chat with the Bazel Python lead though, this is likely to require a change to `rules_python` first. Stay tuned.

**Comment by TimoPtr** [comment]
Do you have any update @gnossen regarding the patch ? We've managed to patch grpc to use the headers from `rules_python`, but out of curiosity we would like to know what is your proposed solution.

**Comment by gowtham-sundara** [comment]
@TimoPtr Could you please provide an example on how you patched it? https://github.com/grpc/grpc/blob/afc040b07a8dcab704fdcab248c8f8b764fafd6c/bazel/grpc_python_deps.bzl#L56 seems to be discovering and binding to a python version from env variables :*

**Comment by adam-azarchs** [comment]
Our team's perhaps equally hacky solution to this problem looks like
```diff
diff --git a/third_party/py/python_configure.bzl b/third_party/py/python_configure.bzl
index fa47d89ebc..364bffbb9a 100644
--- a/third_party/py/python_configure.bzl
+++ b/third_party/py/python_configure.bzl
@@ -119,36 +119,12 @@ def _symlink_genrule_for_dir(
   If src_dir is passed, files will be read from the given directory; otherwise
   we assume files are in src_files and dest_files
   """
-    if src_dir != None:
-        src_dir = _normalize_path(src_dir)
-        dest_dir = _normalize_path(dest_dir)
-        files = "\n".join(
-            sorted(_read_dir(repository_ctx, src_dir).splitlines()),
-        )
-
-        # Create a list with the src_dir stripped to use for outputs.
-        dest_files = files.replace(src_dir, "").splitlines()
-        src_files = files.splitlines()
-    command = []
-    outs = []
-    for i in range(len(dest_files)):
-        if dest_files[i] != "":
-            # If we have only one file to link we do not want to use the dest_dir, as
-            # $(@D) will include the full path to the file.
-            dest = "$(@D)/" + dest_dir + dest_files[i] if len(
-                dest_files,
-            ) != 1 else "$(@D)/" + dest_files[i]
-
-            # On Windows, symlink is not supported, so we just copy all the files.
-            cmd = "cp -f" if _is_windows(repository_ctx) else "ln -s"
-            command.append(cmd + ' "%s" "%s"' % (src_files[i], dest))
-            outs.append('        "' + dest_dir + dest_files[i] + '",')
-    return _genrule(
-        src_dir,
-        genrule_name,
-        " && ".join(command),
-        "\n".join(outs),
-    )
+    return """
+alias(
+    name = "_python3_include",
+    actual = "@python_respoitory//:python_hdrs",
+)
+"""
 
 def _get_python_bin(repository_ctx, bin_path_key, default_bin_path, allow_absent):
     """Gets the python bin path."""
@@ -345,7 +321,7 @@ def _python_autoconf_impl(repository_ctx):
         repository_ctx,
         "_python3",
         _PYTHON3_BIN_PATH,
-        "python3",
+        "../python_repository/bin/python3",
         _PYTHON3_LIB_PATH,
         False
     )
-- 
```

**Comment by gnossen** [comment]
@TimoPtr We cannot rely directly on `rules_python` within the gRPC library since the end user application is in control of the naming of external libraries. The solution here will be to introduce a new toolchain that exposes the Python headers within `rules_python`. This should be completed within the next couple of months.

For the moment, I think the patches proposed in this thread are a fine workaround if you are an application and not a library.

**Comment by aryeh-looker** [comment]
+1 -- fwiw I am running into this on a standard issue gLinux with bazel.

**Comment by chickenandpork** [comment]
@gnossen is it possible to accept one of these workarounds, and then fix it when the 

> Yes. This is something I'm actively working on. After a chat with the Bazel Python lead though, this is likely to require a change to `rules_python` first. Stay tuned.

@gnossen is it possible to accept one of these workarounds in the meantime as a temp solution?  The issue is 2.5yrs old, and the dependent fix in rules_python may take additional time.

**Comment by sfc-gh-kleonhard** [comment]
I'd appreciate having one of the workarounds accepted too!

**Comment by westrustlab** [comment]
We just ran into this issue when upgrading from Python 3.9 to 3.10 - would love to see an update.

**Comment by mering** [comment]
> We just ran into this issue when upgrading from Python 3.9 to 3.10 - would love to see an update.

@westrustlab #35666 is working for us with Bzlmod but it cannot be merged as it breaks WORKSPACE and fails tests. Feel free to suggest a patch to fix the tests. We don't use WORKSPACE ourselves and I also never really used it in the past. So any help appreciated.

**Comment by sfc-gh-dbieliaiev** [comment]
Hi @gnossen. I'd like to help to fix this problem so I tried to revive PR prepared by @mering with fixing a WORKSPACE builds https://github.com/grpc/grpc/pull/40660

I'd appreciate getting some feedback
